subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/capabilities/CapabilitiesUseCasesIntegrationTest.groovy

```diff
@@ -0,0 +1,415 @@
+                            addCapability('org.apache', 'groovy-json', '1.0')
+                        }
+                     }
+                  }
+                  
+                  // solution
+                  configurations.all {
+                      resolutionStrategy {
+                          dependencySubstitution {
+                              substitute module('org.apache:groovy-all') with module('org.apache:groovy-json:1.0')
+                          }
+                      }
+                  }
+               } 
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'org:a:1.0' {
+                expectResolve()
+            }
+            'org:b:1.0' {
+                expectResolve()
+            }
+            'org.apache:groovy:1.0' {
+                expectResolve()
+            }
+            'org.apache:groovy-json:1.0' {
+                expectResolve()
+            }
+        }
+
+        then:
+        run ':checkDeps'
+        resolve.expectGraph {
+            root(":", ":test:") {
+                module('org:a:1.0') {
+                    edge('org.apache:groovy:1.0', 'org.apache:groovy:1.0')
+                    edge('org.apache:groovy-json:1.0', 'org.apache:groovy-json:1.0')
+                }
+                module('org:b:1.0') {
+                    // this is not quite right, as we should replace with 2 edges
+                    // one option to do it is to construct "adhoc" modules, and select an adhoc target in "prefer"
+                    // where this adhoc target would have dependencies on groovy-json and groovy
+                    edge('org.apache:groovy-all:1.0', 'org.apache:groovy-json:1.0')
+                }
+            }
+        }
+    }
+
+    /**
+     * This test highlights the case where published module declares a relocation. This is the drop-in replacement
+     * for "replacedBy" rules when a module has been relocated, and that the publisher uses Gradle. There's the
+     * ability to declare, in a newer version of the module, that it actually provides the same capability as an
+     * older version published at different coordinates.
+     *
+     * This test also makes sure that the order in which dependencies are seen in the graph do not matter.
+     */
+    @RequiredFeatures(
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+    )
+    @Unroll
+    def "published module can declare relocation (first in graph = #first, second in graph = #second)"() {
+        given:
+        repository {
+            'asm:asm:3.0' {
+                variant('runtime') {
+                    capability('asm', 'asm', '3.0')
+                }
+            }
+            'org.ow2.asm:asm:4.0' {
+                variant('runtime') {
+                    capability('asm', 'asm', '4.0') // upgrades the asm capability
+                    capability('org.ow2.asm', 'asm', '4.0') // self capability
```

Comment:
What is the 'self capability' required for?

Reply:
It's not required, it's mostly for documentation purposes, meaning that if you declare a capability that affects a previous (relocated) module, it's likely you want to declare a capability for the newer version.